jsontext/value.go

```diff
@@ -43,70 +58,159 @@ func (v Value) String() string {
+}
+
+// format accepts two []Options to avoid the allocation appending them together.
+// It is equivalent to v.Format(append(opts1, opts2...)...).
+func (v *Value) format(opts1, opts2 []Options) error {
+	e := getBufferedEncoder(opts1...)
+	defer putBufferedEncoder(e)
+	e.s.Join(opts2...)
+	e.s.Flags.Set(jsonflags.OmitTopLevelNewline | 1)
+	if err := e.s.WriteValue(*v); err != nil {
+		return err
+	}
+	if !bytes.Equal(*v, e.s.Buf) {
+		*v = append((*v)[:0], e.s.Buf...)
+	}
+	return nil
+}
+
 // Compact removes all whitespace from the raw JSON value.
 //
-// It does not reformat JSON strings to use any other representation.
-// It is guaranteed to succeed if the input is valid.
-// If the value is already compacted, then the buffer is not mutated.
-func (v *Value) Compact() error {
-	return v.reformat(false, false, "", "")
+// It does not reformat JSON strings or numbers to use any other representation.
+// To maximize the set of JSON values that can be formatted,
+// this permits values with duplicate names and invalid UTF-8.
+//
+// Compact is equivalent to calling [Value.Format] with the following options:
+//   - [AllowDuplicateNames](true)
+//   - [AllowInvalidUTF8](true)
+//   - [PreserveRawStrings](true)
+//
+// Any options specified by the caller are applied after the initial set
+// and may deliberately override prior options.
+func (v *Value) Compact(opts ...Options) error {
+	return v.format([]Options{
+		AllowDuplicateNames(true),
+		AllowInvalidUTF8(true),
+		PreserveRawStrings(true),
+	}, opts)
 }
 
 // Indent reformats the whitespace in the raw JSON value so that each element
-// in a JSON object or array begins on a new, indented line beginning with
-// prefix followed by one or more copies of indent according to the nesting.
-// The value does not begin with the prefix nor any indention,
-// to make it easier to embed inside other formatted JSON data.
+// in a JSON object or array begins on a indented line according to the nesting.
+//
+// It does not reformat JSON strings or numbers to use any other representation.
+// To maximize the set of JSON values that can be formatted,
+// this permits values with duplicate names and invalid UTF-8.
 //
-// It does not reformat JSON strings to use any other representation.
-// It is guaranteed to succeed if the input is valid.
-// If the value is already indented properly, then the buffer is not mutated.
+// Indent is equivalent to calling [Value.Format] with the following options:
+//   - [AllowDuplicateNames](true)
+//   - [AllowInvalidUTF8](true)
+//   - [PreserveRawStrings](true)
+//   - [Multiline](true)
 //
-// The prefix and indent strings must be composed of only spaces and/or tabs.
-func (v *Value) Indent(prefix, indent string) error {
-	return v.reformat(false, true, prefix, indent)
+// Any options specified by the caller are applied after the initial set
+// and may deliberately override prior options.
+func (v *Value) Indent(opts ...Options) error {
```

Comment:
This change is so good 👌🏻

Reply:
You can thank @neild. When he reviewed https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#MarshalOptions.Multiline, he advocated for a single `Multiline` bool that would choose a default indent because he didn't want to think about what it should be. This is carrying forth the same philosophy.